File: 0110-balanced-binary-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    if root == nil {
        return true
    }

    balanced, _ := isBalancedHelper(root)
    return balanced
}

func isBalancedHelper(root *TreeNode) (bool, int) {
    if root == nil {
        return true, 0
    }

    isbalancedleft, left := isBalancedHelper(root.Left)
    isbalancedright, right := isBalancedHelper(root.Right)
    absDiff := left - right
    if absDiff < 0 {
        absDiff = -1 * absDiff
    }

    return isbalancedleft && isbalancedright && absDiff < 2 , max(left + 1, right + 1)
}

func max(a, b int) int {
    if a > b {
        return a
    }

    return b
}

File: 0110-balanced-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def helper(root):
            if root is None:
                return (True, 0)
            
            leftDepth = helper(root.left)
            if leftDepth[0] is False:
                return (False, leftDepth[1])

            rightDepth = helper(root.right)
            if rightDepth[0] is False:
                return (False, rightDepth[1])

            if abs(leftDepth[1] - rightDepth[1]) > 1:
                return (False, 0)

            return (True, 1 + max(leftDepth[1], rightDepth[1]))

        return helper(root)[0]
        
